### PR TITLE
feature: Add quantum algorithms

### PR DESCRIPTION
**Pull request by math411**
*Issue #, if available:*

*Description of changes:*
feature: Add quantum algorithms
*Testing done:*
tox
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ x ] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [ x ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ x ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ x ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ x ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/circuits/quantum_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+# language governing permissions and limitations under the License.
+
+from __future__ import annotations
+
+from typing import List
+
+from braket.circuits.quantum_operator import QuantumOperator
+
+
+class QuantumAlgorithm(QuantumOperator):
```
Comment (kshitijc) [inline src/braket/circuits/quantum_algorithm.py:21]: I'm not sure if `QuantumAlgorithm` is what this represents. The docstring seems to indicate that this behaves more like a subcircuit.

File: src/braket/circuits/quantum_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        Examples:
+            >>> circ = Circuit().qft([0,1,2])
+        """
+        return Instruction(QFT(targets), target=targets)
+
+
+QuantumAlgorithm.register_quantum_algorithm(QFT)
+
+
+def generate_algo_ascii(algo_name: str, qubit_count: int) -> Sequence[str]:
```
Comment (kshitijc) [inline src/braket/circuits/quantum_algorithms.py:96]: Could we add a test to show what this produces ?

File: src/braket/circuits/circuit.py [hunk 3]
```diff
@@ -710,13 +710,15 @@ def apply_gate_noise(
         if not self.qubits:
             raise IndexError("Gate noise cannot be applied to an empty circuit.")
 
+        decomposed_circ = self.decompose()
```
Comment (kshitijc) [inline src/braket/circuits/circuit.py:713]: Why should this always decompose on application of noise ?

File: src/braket/circuits/quantum_algorithm.py (hunk 1 above)
Comment (math411) [reply src/braket/circuits/quantum_algorithm.py:21]: That is a typo. It should read as follows:
```
An algorithm that can be distinctly represented in a circuit.
```

File: src/braket/circuits/circuit.py (hunk 3 above)
Comment (math411) [reply src/braket/circuits/circuit.py:713]: Decomposition can also happen at execution time.
This allows for seeing what gates have noise applied and doesn't require tracking noise to be applied later to the QuantAlgo.

**Pull request closed without merging**